azavea/raster-foundry
# Use SWF for async tasks/workflows

This commit adds support for using SWF to run asynchronous tasks and
workflows using the botoflow library. This commit includes stubbed out
functionality for handling and kicking off what could be basic tasks for
processing image uploads.

The `click` library is used for building a command line client for
starting activity and decider workers that process jobs as they come
in.

An SWF domain is created for developers for development and a single
activity and decider worker is started with docker compose.

## Testing
- follow setup instructions (ensuring an AWS profile for `raster-foundry` is configured)
- run `vagrant provision`
- ssh into vagrant machine
- start server with `./scripts/server` and wait for services to start
- once services are started, in another console kick off an SWF job with `./scripts/console app-tasks-decider "rf process_image abc.tif"` and observer output
- log in to AWS console and browse to the SWF dashboard, select your domain, and check to see that the job was marked as a success

## Notes
We've used SWF in the past with mixed success. While SWF has many desirable features, the developer and debugging experience has been less than stellar for some of the following reasons.
 - required to use an unfamiliar language (`ruby` or `java`)
 - poor console/admin interface

However, there are many attractive features that suggest it would be an appropriate choice for Raster Foundry that include the following:
 - a new, python based framework for SWF under the `boto` project
 - better CloudWatch metrics and dashboards that could make it easier
 - the ability to version jobs to make blue-green deployments easier
 - ability to pin jobs to a particular machine by using task lists

There are some additional risks associated with SWF or things we'll need to deal with:
 - `botoflow` is not fully stable yet, though the maintainer doesn't see any major refactors coming up
 - we would have to come up with a better way to view running jobs, potentially replacing some parts of the AWS console with our own console.

This PR I think demonstrates what we would want as a bare minimum for _any_ framework -- the ability to start up workers to process jobs that are kicked off simultaneously. The fact that it is in python, follows an execution model we already know, and avoids the task of figuring out how to maintain state for workflows makes this particularly attractive as well.
